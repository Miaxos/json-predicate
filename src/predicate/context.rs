use std::error::Error;

use crate::json_path::{JSONPath, JSONPathError};

/// The context to run the evaluation, you can have an empty context with
/// `PredicateContext::default()`.
#[derive(Default, Clone, Debug)]
pub struct PredicateContext {
    location: Option<JSONPath>,
}

#[derive(Debug, thiserror::Error)]
pub enum PredicateContextError {
    #[error("{0}")]
    Deserialize(#[from] Box<dyn Error>),
    #[error("{0}")]
    JSONPath(#[from] JSONPathError),
}

impl From<JSONPath> for PredicateContext {
    fn from(value: JSONPath) -> Self {
        Self {
            location: Some(value),
        }
    }
}

impl From<Option<JSONPath>> for PredicateContext {
    fn from(value: Option<JSONPath>) -> Self {
        Self { location: value }
    }
}

impl PredicateContext {
    pub fn new(path: String) -> Result<Self, PredicateContextError> {
        let location = JSONPath::new(path)?;

        Ok(Self {
            location: Some(location),
        })
    }

    pub fn final_path(&self, path: &Option<JSONPath>) -> Option<JSONPath> {
        // TODO: Remove cloning
        match (self.location.clone(), path.clone()) {
            (None, None) => None,
            (Some(path), None) | (None, Some(path)) => Some(path),
            (Some(root), Some(path)) => Some(root + path),
        }
    }
}
