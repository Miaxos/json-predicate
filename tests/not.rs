// Test adapted from https://github.com/MalcolmDwyer/json-predicate/blob/master/test/test.js

use json_predicate::builder::NotBuilder;
use json_predicate::context::PredicateContext;
use json_predicate::{Predicate, PredicateImpl, SecondOrder};

mod utils;
use serde::Deserialize;
use utils::ENTRY;

#[test]
pub fn test_base_predicate_snapshot() {
    let predicate: Predicate =
        SecondOrder::from(NotBuilder::default().apply([]).build().unwrap()).into();

    insta::assert_json_snapshot!(predicate);
}

#[test]
pub fn return_true_for_not_case_f_f_with_shallow_endpoint() {
    let predicate = Predicate::deserialize(serde_json::json!({
        "op": "not",
        "apply": [
          {
            "op": "defined",
            "path": "/not_real_thing"
          },
          {
            "op": "defined",
            "path": "/not_real_1"
          }
        ],
    }));

    assert!(predicate.is_ok());
    let result = predicate
        .unwrap()
        .evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert!(result.unwrap());
}

#[test]
pub fn return_true_for_not_case_f_f_with_deep_endpoint() {
    let predicate = Predicate::deserialize(serde_json::json!({
        "op": "not",
        "apply": [
          {
            "op": "defined",
            "path": "/objA/not_real_thing"
          },
          {
            "op": "defined",
            "path": "/objA/not_real_1"
          }
        ],
    }));

    assert!(predicate.is_ok());
    let result = predicate
        .unwrap()
        .evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert!(result.unwrap());
}

#[test]
pub fn return_false_for_not_case_t_f_and_compound_paths() {
    let predicate = Predicate::deserialize(serde_json::json!({
        "op": "not",
        "path": "/objA",
        "apply": [
          {
            "op": "defined",
            "path": "/not_real_thing"
          },
          {
            "op": "defined",
            "path": "/stringXYZ"
          }
        ],
    }));

    assert!(predicate.is_ok());
    let result = predicate
        .unwrap()
        .evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert!(!result.unwrap());
}

#[test]
pub fn return_false_for_or_case_t_f_t() {
    let predicate = Predicate::deserialize(serde_json::json!({
        "op": "not",
        "path": "/objA",
        "apply": [
          {
            "op": "defined",
            "path": "/stringX"
          },
          {
            "op": "defined",
            "path": "/not_real"
          },
          {
            "op": "defined",
            "path": "/null2"
          }
        ],
    }));

    assert!(predicate.is_ok());
    let result = predicate
        .unwrap()
        .evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert!(!result.unwrap());
}

#[test]
pub fn return_true_for_not_case_f_f() {
    let predicate = Predicate::deserialize(serde_json::json!({
        "op": "not",
        "path": "/objA",
        "apply": [
          {
            "op": "defined",
            "path": "/not_real_1"
          },
          {
            "op": "defined",
            "path": "/not_real_2"
          },
        ],
    }));

    assert!(predicate.is_ok());
    let result = predicate
        .unwrap()
        .evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert!(result.unwrap());
}

#[test]
pub fn returns_true_for_not_case_empty_array() {
    let predicate = Predicate::deserialize(serde_json::json!({
        "op": "not",
        "path": "/objA",
        "apply": [],
    }));

    assert!(predicate.is_ok());
    let result = predicate
        .unwrap()
        .evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert!(result.unwrap());
}
