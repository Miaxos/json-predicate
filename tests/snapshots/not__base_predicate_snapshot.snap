---
source: tests/not.rs
expression: predicate

---
{
  "op": "not",
  "path": null,
  "apply": []
}
